plugin/src/main/java/com/google/errorprone/xplat/checker/J2objcMethodName.java

```diff
@@ -0,0 +1,351 @@
+        }
+      }
+    }
+    output.append("_");
+    return output.toString();
+  }
+
+  private String methodNameMangle(MethodTree tree, VisitorState state) {
+    StringBuilder output = new StringBuilder();
+
+    Symbol symbol = ASTHelpers.getSymbol(tree);
+
+    if (foundObjcClassNames.containsKey(symbol.enclClass().getSimpleName().toString())) {
+      output.append(foundObjcClassNames.get(symbol.enclClass().getSimpleName().toString()));
+    } else {
+      output.append(
+          localClassNameMangle(symbol.enclClass().toString(), symbol.outermostClass().toString()));
+    }
+
+    output.append("_");
+
+    if (!tree.getModifiers().getAnnotations().isEmpty()) {
+      for (AnnotationTree annTree : tree.getModifiers().getAnnotations()) {
+        String source = state.getSourceForNode(annTree.getAnnotationType());
+
+        if (source != null && source.equals("ObjectiveCName")) {
+          String value = annTree.getArguments().toString();
+          output.append(value.substring(value.indexOf("\"") + 1, value.lastIndexOf("\"")));
+          break;
+        }
+      }
+    } else {
+      if (tree.getParameters().isEmpty()) {
+        output.append(symbol.name);
+      } else {
+        output.append(parameterNameMangle(symbol.name.toString(), tree.getParameters()));
+      }
+    }
+    return output.toString();
+  }
+
+  private String getMethodName(String methodName) {
+    if (usedMethodNames.containsKey(methodName)) {
+      usedMethodNames.put(methodName, usedMethodNames.get(methodName) + 1);
+      return String.format("%s%d", methodName, usedMethodNames.get(methodName));
+    } else {
+      usedMethodNames.put(methodName, 1);
+      return methodName;
+    }
+  }
+
+  public J2objcMethodName(ErrorProneFlags flags) {
+
+    Optional<Integer> arg = flags.getInteger("J2ObjCMethodName:MethodNameLength");
+
+    if (arg.isPresent()) {
+      this.methodNameLength = arg.get();
+    }
+  }
+
+  @Override
+  public Description matchMethod(MethodTree tree, VisitorState state) {
+    if (MATCHER.matches(tree, state)) {
+      String mangledName = methodNameMangle(tree, state);
+
+      if (mangledName.length() >= 100 && tree.getParameters().size() > 5) {
+        return genDescription(
+            String.format("This method should likely be refactored to have fewer"
+                + " parameters and its name will be %d characters when translated to"
+                + " Objective-C: %s", mangledName.length(), mangledName), tree);
+
+      } else if (mangledName.length() >= methodNameLength) {
+        return genDescription(
+            String.format("This method name will be %d characters when translated to"
+                + " Objective-C: %s", mangledName.length(), mangledName), tree);
+      }
+    }
+
+    return Description.NO_MATCH;
+  }
+
+  /**
+   * Looks for an ObjectiveCName annotation on a class. If one is found, stores the class name as
```

Comment:
So basically, this method and `matchCompilationUnit` are both needed to build the foundObjcClassNames (this one) or to set a value to `this.packageAnnotation`, and are not checks themselves, right? Perhaps explain that in the Javadoc here and in the next method.

Reply:
Yes, that is correct. I have made that more clear in the Javadoc comment.